Use terraform to install the masters

markmc opened the pull request:
Install terraform and terraform-provider-ironic. Add a script that you can use to experiment with using terraform to deploy the masters after 07_deploy_masters.sh exits early.

This uses the unmerged code in metalkube/terraform-provider-ironic#2

---

TODO list before merging:

- [x] - Merge metalkube/terraform-provider-ironic#2
- [x] - Implement "destroy" in terraform-provider-ironic
- [x] - Fix how terraform-provider-ironic is built, including pulling in latest gophercloud/utils
- [x] - Use "terraform destroy" in masters_cleanup.sh

hardys commented on 03_ocp_repo_sync.sh:
```diff
@@ -49,3 +49,23 @@ pushd "${GOPATH}/src/github.com/metalkube/facet"
+git reset --hard pr/82
+
+cd $GOPATH/src/github.com/metalkube/terraform-provider-ironic
+git fetch origin +refs/pull/2/head:remotes/pr/2
```
I wonder if we should modify sync_go_repo_and_patch so it can do the fetch/checkout instead of the curl/git am?

Seems like consuming in-progress PRs is likely the common pattern, so we could clean things up a bit that way perhaps?

markmc replied:
Yeah, honestly all I was doing here was hacking some way of building this with @stbenjam's gophercloud/utils PR. We should just merge the terraform-provider-ironic PR

stbenjam replied:
I think that's fine, I can continue development in additional PR's.  I just wanted to make sure there's nothing missing for us in the MVP of the provider.  I did just push a small fix for an error with it not recognizing "wait call-back" as an intermediary step in deployment, but otherwise I think it should be good to go for now.

stbenjam replied:
I'll also push the changes to `gophercloud/utils` they asked for, and hope that should get merged pretty quickly.

markmc replied:
+1 for that plan

The "wait call-back" change helped, thanks!

hardys commented:
Nice I was about to do something very similar to this :)

stbenjam commented on 07_deploy_masters.sh:
```diff
@@ -29,6 +29,9 @@ for node in $(jq -r .nodes[].name ${MASTER_NODES_FILE}); do
 done
 
+echo "Have fun playing with test-terraform.sh !"
+exit 0
```
Could we just directly run the terraform script from here, maybe gated by an env variable?

stbenjam replied:
Also the script is "terraform-test.sh"

hardys replied:
I'd say lets just run terraform here, as a step towards running it via kni-install?

markmc replied:
Totally, shouldn't take much more work to just have 07_deploy_masters.sh use terraform - I just pushed this as WIP for playing with

markmc replied:
Ok, switched over to terraform here - it's kicking off the node deployments ok, but I suspect the igntion config isn't getting through - haven't debugged yet

stbenjam replied:
Ah, yea, it looks like it's not getting through, I think it might be a bug in the provider, looking into it

stbenjam replied:
So, it was my misunderstanding of how `DiffSurpressFunc` worked in Terraform.  It should work now, or at least it actually builds the config drive.

stbenjam commented:
This looks great!

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/149/

stbenjam commented on 01_install_requirements.sh:
```diff
@@ -62,3 +62,11 @@ fi
+# Install terraform
+if [ ! -f /usr/local/bin/terraform ]; then
+    curl -O https://releases.hashicorp.com/terraform/0.11.11/terraform_0.11.11_linux_amd64.zip
+    unzip terraform_0.11.11_linux_amd64.zip
```
Could you install unzip as well? My CentOS 7 box from beaker didn't have it.

markmc replied:
Thanks for pointing that out

You're good with installing terraform this way vs via an RPM or "go get"? I didn't look into options

stbenjam replied:
I don't believe they produce any official RPM's, this is exactly how I installed it.

markmc replied:
Added unzip, thanks

stbenjam commented on 07_deploy_masters.sh:
```diff
@@ -20,45 +20,33 @@ if [ $(sudo podman ps | grep -w -e "ironic$" -e "ironic-inspector$" | wc -l) !=
+echo "Deploying master nodes"
+cd ocp/tf-master
+terraform init
+terraform apply --auto-approve
```
Could we enable debug logging as well? 

`export TF_LOG=DEBUG`

markmc replied:
Good call!

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/156/

markmc commented:
I rebased and squashed. Masters seem to deploy fine to me.

TODO list before merging:

- [ ] - Merge metalkube/terraform-provider-ironic#2
- [ ] - Implement "destroy" in terraform-provider-ironic
- [ ] - Fix how terraform-provider-ironic is built, including pulling in latest gophercloud/utils
- [ ] - Use "terraform destroy" in masters_cleanup.sh

Anything I missed?

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/174/

stbenjam commented:
I think we could defer destroy to a different JIRA ticket, if we want to get this in sooner rather than later.

Do we want to wait for https://github.com/gophercloud/utils/pull/82? We could just leave the hack to pull from the PR until it's merged.

stbenjam commented:
Actually deleted is easy enough to implement, it's in the terraform provider now.

markmc commented:
> Actually deleted is easy enough to implement, it's in the terraform provider now.

Awesome, I'll try it out.

> Do we want to wait for [gophercloud/utils#82](https://github.com/gophercloud/utils/pull/82)? We could just leave the hack to pull from the PR until it's merged.

Surely we can figure out some way to make the terraform-provider-ironic build from a clean checkout without having to modify a version-controlled file? Can't we vendor your fork? I hadn't looked at go modules too closely before but ... you'd hope this wouldn't be such a crazy thing to want to do

markmc commented:
> > Actually deleted is easy enough to implement, it's in the terraform provider now.
> 
> Awesome, I'll try it out.

Looks to be just about working. Got the error below first eim, then re-ran it, and it completed with no errors:

```
+ terraform destroy --auto-approve
ironic_node_v1.openshift-master-0: Refreshing state... (ID: 9f0877f5-e47d-4337-acdd-bf9e753f1181)
ironic_node_v1.openshift-master-2: Refreshing state... (ID: a202efa7-ed06-4581-bddb-a0d5b10419f5)
ironic_node_v1.openshift-master-1: Refreshing state... (ID: 7ec4a692-e4e4-47e1-9f6e-37295d61e15d)
ironic_node_v1.openshift-master-0: Destroying... (ID: 9f0877f5-e47d-4337-acdd-bf9e753f1181)
ironic_node_v1.openshift-master-2: Destroying... (ID: a202efa7-ed06-4581-bddb-a0d5b10419f5)
ironic_node_v1.openshift-master-1: Destroying... (ID: 7ec4a692-e4e4-47e1-9f6e-37295d61e15d)

Error: Error applying plan:

3 error(s) occurred:

* ironic_node_v1.openshift-master-0 (destroy): 1 error(s) occurred:

* ironic_node_v1.openshift-master-0: cannot delete node in state 'deleting'
* ironic_node_v1.openshift-master-1 (destroy): 1 error(s) occurred:

* ironic_node_v1.openshift-master-1: cannot delete node in state 'deleting'
* ironic_node_v1.openshift-master-2 (destroy): 1 error(s) occurred:

* ironic_node_v1.openshift-master-2: cannot delete node in state 'deleting'
```

stbenjam commented:
That should be fixed now, I read Ironic's state diagram wrong.

I'll have a look about vendoring my gohpercloud changes until the PR gets merged.  https://github.com/Masterminds/glide looks promising.

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/181/

stbenjam commented:
I updated https://github.com/metalkube/terraform-provider-ironic/pull/2 to use `glide` which let me vendor my fork of `gophercloud\utils`. Unfortunately glide's dep management is a bit more aggressive than the go.mod stuff, it pulled in a lot more things from terraform... Once we're less reliant on forks we could maybe move back to go's native mod handling.

Do you want to try removing the hack for utils? Then we can merge https://github.com/metalkube/terraform-provider-ironic/pull/2

markmc commented:
> I updated [metalkube/terraform-provider-ironic#2](https://github.com/metalkube/terraform-provider-ironic/pull/2) to use `glide` which let me vendor my fork of `gophercloud/utils`. Unfortunately glide's dep management is a bit more aggressive than the go.mod stuff, it pulled in a lot more things from terraform... Once we're less reliant on forks we could maybe move back to go's native mod handling.

Yeah, it's quite a thing! Agree we can move back later

> Do you want to try removing the hack for utils? Then we can merge [metalkube/terraform-provider-ironic#2](https://github.com/metalkube/terraform-provider-ironic/pull/2)

Seems to be working for me. Thanks!

stbenjam commented:
@markmc I merged https://github.com/metalkube/terraform-provider-ironic/pull/2, you can remove the hack for that too - and we should be good to go!

markmc commented:
Ok, I think this is good to go if CI passes. I'm re-testing locally also.

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/188/

markmc commented:
> Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/188/

This is a result on the earlier push, where we were still using metalkube/terraform-provider-ironic#2

markmc commented:
> Ok, I think this is good to go if CI passes. I'm re-testing locally also.

Looks good from my local testing

derekhiggins commented:
Build SUCCESS, see build http://10.8.144.11:8080/job/dev-tools/189/

hardys commented:
Looks like we need a rebase but I can pull/test when done if we think this is ready to go?

Otherwise lgtm - seems like a great step towards driving the master deployment via kni-installer :)

markmc commented:
Go for it Steve, thanks!

markmc commented:
Rebased but haven't tested yet

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/225/

markmc commented:
Ok, works fine for me with openshift-metalkube#165

hardys commented:
@stbenjam Hey did the concerns about the Ironic API load/locking get resolved, or should we hold this pending further investigation?

stbenjam commented:
@hardys It's not resolved. Throughout testing I've deployed with terraform maybe a dozen times and I've personally run into the sqlite3 locking issue twice. I have only seen it while I'm deploying and doing something like `watch -n10 openstack baremetal node list` at the same time.

@markmc Have you seen it at all?

derekhiggins commented:
Build FAILURE, see build http://10.8.144.11:8080/job/dev-tools/228/

markmc commented:
With this PR, Yurii is seeing:

```
OperationalError: (sqlite3.OperationalError) database is locked [SQL: u'SELECT nodes.created_at
```

even with:

```
$ podman exec -it  ironic sqlite3 /var/lib/ironic/ironic.db  “PRAGMA journal_mode”
wal
```

Dmitri suggests:

> So, our last resort (?) option with sqlite is to try setting `busy_timeout`.

markmc commented:
> @markmc Have you seen it at all?

No. I haven't.

hardys commented:
Ok tested and this works for me, lets merge it and iterate on the sqlite thing if it recreates.

One thing to note is folks need to destroy with the old cleanup before consuming this patch, as (understandably) the new destroy script fails due to missing ocp/tf-master

yprokule commented:
> With this PR, Yurii is seeing:
> 
> ```
> OperationalError: (sqlite3.OperationalError) database is locked [SQL: u'SELECT nodes.created_at
> ```
> 
> even with:
> 
> ```
> $ podman exec -it  ironic sqlite3 /var/lib/ironic/ironic.db  “PRAGMA journal_mode”
> wal
> ```
> 
> Dmitri suggests:
> 
> > So, our last resort (?) option with sqlite is to try setting `busy_timeout`.

Failed like:
```bash
ironic_node_v1.openshift-master-2: Still creating... (2m50s elapsed)
ironic_node_v1.openshift-master-0: Still creating... (2m50s elapsed)
ironic_node_v1.openshift-master-1: Still creating... (2m50s elapsed)
2019/03/15 10:29:45 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* ironic_node_v1.openshift-master-1: Internal Server Error
2019/03/15 10:29:45 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* ironic_node_v1.openshift-master-1: Internal Server Error
2019/03/15 10:29:45 [TRACE] [walkApply] Exiting eval tree: ironic_node_v1.openshift-master-1
2019/03/15 10:29:48 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* ironic_node_v1.openshift-master-2: Internal Server Error
2019/03/15 10:29:48 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* ironic_node_v1.openshift-master-2: Internal Server Error

2019/03/15 10:30:02 [DEBUG] plugin: waiting for all plugin processes to complete...
Error: Error applying plan:

3 error(s) occurred:

* ironic_node_v1.openshift-master-0: 1 error(s) occurred:

* ironic_node_v1.openshift-master-0: Internal Server Error
* ironic_node_v1.openshift-master-1: 1 error(s) occurred:

* ironic_node_v1.openshift-master-1: Internal Server Error
* ironic_node_v1.openshift-master-2: 1 error(s) occurred:

2019-03-15T10:30:02.972+0200 [DEBUG] plugin.terraform-provider-ironic: 2019/03/15 10:30:02 [ERR] plugin: stream copy 'stderr' error: stream closed
* ironic_node_v1.openshift-master-2: Internal Server Error

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.


2019-03-15T10:30:03.007+0200 [DEBUG] plugin.terraform-provider-ironic: 2019/03/15 10:30:03 [ERR] plugin: plugin server: accept unix /tmp/plugin005349256: use of closed network connection
2019-03-15T10:30:03.008+0200 [DEBUG] plugin: plugin process exited: path=/root/.terraform.d/plugins/terraform-provider-ironic
```
```bash
2019-03-15 08:29:25.649 44 ERROR wsme.api [req-747fc5e4-6050-463e-9d5d-8b7fa79a00f3 - - - - -] Server-side error: "(sqlite3.OperationalError) database is locked [SQL: u'SELECT anon_1.nodes_created_at AS anon_1_nodes_created_at, anon_1.nodes_updated_at AS anon_1_nodes_updated_at, anon_1.nodes_version AS anon_1_nodes_version, anon_1.nodes_id AS anon_1_nodes_id, anon_1.nodes_uuid AS anon_1_nodes_uuid, anon_1.nodes_instance_uuid AS anon_1_nodes_instance_uuid, anon_1.nodes_name AS anon_1_nodes_name, anon_1.nodes_chassis_id AS anon_1_nodes_chassis_id, anon_1.nodes_power_state AS anon_1_nodes_power_state, anon_1.nodes_target_power_state AS anon_1_nodes_target_power_state, anon_1.nodes_provision_state AS anon_1_nodes_provision_state, anon_1.nodes_target_provision_state AS anon_1_nodes_target_provision_state, anon_1.nodes_provision_updated_at AS anon_1_nodes_provision_updated_at, anon_1.nodes_last_error AS anon_1_nodes_last_error, anon_1.nodes_instance_info AS anon_1_nodes_instance_info, anon_1.nodes_properties AS anon_1_nodes_properties, anon_1.nodes_driver AS anon_1_nodes_driver, anon_1.nodes_driver_info AS anon_1_nodes_driver_info, anon_1.nodes_driver_internal_info AS anon_1_nodes_driver_internal_info, anon_1.nodes_clean_step AS anon_1_nodes_clean_step, anon_1.nodes_deploy_step AS anon_1_nodes_deploy_step, anon_1.nodes_resource_class AS anon_1_nodes_resource_class, anon_1.nodes_raid_config AS anon_1_nodes_raid_config, anon_1.nodes_target_raid_config AS anon_1_nodes_target_raid_config, anon_1.nodes_reservation AS anon_1_nodes_reservation, anon_1.nodes_conductor_affinity AS anon_1_nodes_conductor_affinity, anon_1.nodes_conductor_group AS anon_1_nodes_conductor_group, anon_1.nodes_maintenance AS anon_1_nodes_maintenance, anon_1.nodes_maintenance_reason AS anon_1_nodes_maintenance_reason, anon_1.nodes_fault AS anon_1_nodes_fault, anon_1.nodes_console_enabled AS anon_1_nodes_console_enabled, anon_1.nodes_inspection_finished_at AS anon_1_nodes_inspection_finished_at, anon_1.nodes_inspection_started_at AS anon_1_nodes_inspection_started_at, anon_1.nodes_extra AS anon_1_nodes_extra, anon_1.nodes_automated_clean AS anon_1_nodes_automated_clean, anon_1.nodes_protected AS anon_1_nodes_protected, anon_1.nodes_protected_reason AS anon_1_nodes_protected_reason, anon_1.nodes_owner AS anon_1_nodes_owner, anon_1.nodes_allocation_id AS anon_1_nodes_allocation_id, anon_1.nodes_description AS anon_1_nodes_description, anon_1.nodes_bios_interface AS anon_1_nodes_bios_interface, anon_1.nodes_boot_interface AS anon_1_nodes_boot_interface, anon_1.nodes_console_interface AS anon_1_nodes_console_interface, anon_1.nodes_deploy_interface AS anon_1_nodes_deploy_interface, anon_1.nodes_inspect_interface AS anon_1_nodes_inspect_interface, anon_1.nodes_management_interface AS anon_1_nodes_management_interface, anon_1.nodes_network_interface AS anon_1_nodes_network_interface, anon_1.nodes_raid_interface AS anon_1_nodes_raid_interface, anon_1.nodes_rescue_interface AS anon_1_nodes_rescue_interface, anon_1.nodes_storage_interface AS anon_1_nodes_storage_interface, anon_1.nodes_power_interface AS anon_1_nodes_power_interface, anon_1.nodes_vendor_interface AS anon_1_nodes_vendor_interface, node_traits_1.created_at AS node_traits_1_created_at, node_traits_1.updated_at AS node_traits_1_updated_at, node_traits_1.version AS node_traits_1_version, node_traits_1.node_id AS node_traits_1_node_id, node_traits_1.trait AS node_traits_1_trait, node_tags_1.created_at AS node_tags_1_created_at, node_tags_1.updated_at AS node_tags_1_updated_at, node_tags_1.version AS node_tags_1_version, node_tags_1.node_id AS node_tags_1_node_id, node_tags_1.tag AS node_tags_1_tag \nFROM (SELECT nodes.created_at AS nodes_created_at, nodes.updated_at AS nodes_updated_at, nodes.version AS nodes_version, nodes.id AS nodes_id, nodes.uuid AS nodes_uuid, nodes.instance_uuid AS nodes_instance_uuid, nodes.name AS nodes_name, nodes.chassis_id AS nodes_chassis_id, nodes.power_state AS nodes_power_state, nodes.target_power_state AS nodes_target_power_state, nodes.provision_state AS nodes_provision_state, nodes.target_provision_state AS nodes_target_provision_state, nodes.provision_updated_at AS nodes_provision_updated_at, nodes.last_error AS nodes_last_error, nodes.instance_info AS nodes_instance_info, nodes.properties AS nodes_properties, nodes.driver AS nodes_driver, nodes.driver_info AS nodes_driver_info, nodes.driver_internal_info AS nodes_driver_internal_info, nodes.clean_step AS nodes_clean_step, nodes.deploy_step AS nodes_deploy_step, nodes.resource_class AS nodes_resource_class, nodes.raid_config AS nodes_raid_config, nodes.target_raid_config AS nodes_target_raid_config, nodes.reservation AS nodes_reservation, nodes.conductor_affinity AS nodes_conductor_affinity, nodes.conductor_group AS nodes_conductor_group, nodes.maintenance AS nodes_maintenance, nodes.maintenance_reason AS nodes_maintenance_reason, nodes.fault AS nodes_fault, nodes.console_enabled AS nodes_console_enabled, nodes.inspection_finished_at AS nodes_inspection_finished_at, nodes.inspection_started_at AS nodes_inspection_started_at, nodes.extra AS nodes_extra, nodes.automated_clean AS nodes_automated_clean, nodes.protected AS nodes_protected, nodes.protected_reason AS nodes_protected_reason, nodes.owner AS nodes_owner, nodes.allocation_id AS nodes_allocation_id, nodes.description AS nodes_description, nodes.bios_interface AS nodes_bios_interface, nodes.boot_interface AS nodes_boot_interface, nodes.console_interface AS nodes_console_interface, nodes.deploy_interface AS nodes_deploy_interface, nodes.inspect_interface AS nodes_inspect_interface, nodes.management_interface AS nodes_management_interface, nodes.network_interface AS nodes_network_interface, nodes.raid_interface AS nodes_raid_interface, nodes.rescue_interface AS nodes_rescue_interface, nodes.storage_interface AS nodes_storage_interface, nodes.power_interface AS nodes_power_interface, nodes.vendor_interface AS nodes_vendor_interface \nFROM nodes ORDER BY nodes.id ASC\n LIMIT ? OFFSET ?) AS anon_1 LEFT OUTER JOIN node_traits AS node_traits_1 ON node_traits_1.node_id = anon_1.nodes_id LEFT OUTER JOIN node_tags AS node_tags_1 ON node_tags_1.node_id = anon_1.nodes_id ORDER BY anon_1.nodes_id ASC'] [parameters: (1000, 0)] (Background on this error at: http://sqlalche.me/e/e3q8)". Detail:
Traceback (most recent call last):

  File "/usr/lib/python2.7/site-packages/wsmeext/pecan.py", line 85, in callfunction
    result = f(self, *args, **kwargs)

  File "/usr/lib/python2.7/site-packages/ironic/api/controllers/v1/node.py", line 1872, in get_all
    **extra_args)

  File "/usr/lib/python2.7/site-packages/ironic/api/controllers/v1/node.py", line 1684, in _get_nodes_collection
    filters=filters)

  File "/usr/lib/python2.7/site-packages/ironic/objects/node.py", line 313, in list
    sort_dir=sort_dir)

  File "/usr/lib/python2.7/site-packages/ironic/db/sqlalchemy/api.py", line 400, in get_node_list
    sort_key, sort_dir, query)

  File "/usr/lib/python2.7/site-packages/ironic/db/sqlalchemy/api.py", line 229, in _paginate_query
    return query.all()

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/query.py", line 2925, in all
    return list(self)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/query.py", line 3081, in __iter__
    return self._execute_and_instances(context)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/orm/query.py", line 3106, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 980, in execute
    return meth(self, multiparams, params)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/sql/elements.py", line 273, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1099, in _execute_clauseelement
    distilled_params,

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1240, in _execute_context
    e, statement, parameters, cursor, context

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1456, in _handle_dbapi_exception
    util.raise_from_cause(newraise, exc_info)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/util/compat.py", line 296, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/base.py", line 1236, in _execute_context
    cursor, statement, parameters, context

  File "/usr/lib64/python2.7/site-packages/sqlalchemy/engine/default.py", line 536, in do_execute
    cursor.execute(statement, parameters)
```
